Add pre-commit config and CI job

jpthiele opened the pull request:
As proposed in #17860, this adds a `.pre-commit-config.yaml` to configure the tool [pre-commit](https://pre-commit.com/) with a few hooks that seemed sensible to me.

I have added the two hooks that did not pass by default as separate commits so they could be removed more easily if they should not be wanted.

I also added a workflow file that runs all pre-commit checks except clang-format on each PR. 
In theory this could make the indent CI obsolete, but I know that it does more than run clang-format.

jpthiele commented:
One potential hook we could add later is for `codespell`,
which is run manually for each release. 
We would need to write a `.codespellrc` and teach it to ignore some false positive errors, e.g.
Numer.,OCE,preCICE.
Pinging @marcfehling about this because you ran codespell on 9.6 and probably have some insights into what to ignore.

marcfehling commented:
> Pinging @marcfehling about this because you ran codespell on 9.6 and probably have some insights into what to ignore.

We are running [`typos` as part of our CI](https://github.com/dealii/dealii/blob/16484460f0d2a2a02dcb45acb31b2715277dc3cf/.github/workflows/indent.yml#L72-L83), and have a corresponding configuration file [`.typos.toml`](https://github.com/dealii/dealii/blob/master/.typos.toml) in the root folder of the repository.

Would it be possible to add a hook for `typos` instead of `codespell` instead?

marcfehling reviewed (approved):
I like this! :+1:

Could you please also add a changelog entry for this new feature?

And please also add some instructions on how to set-up `pre-commit` with all its dependencies. I guess a new entry in this [Wiki section](https://github.com/dealii/dealii/wiki#developingcontributing) would be a good place.

jpthiele commented:
> > Pinging @marcfehling about this because you ran codespell on 9.6 and probably have some insights into what to ignore.
> 
> We are running [`typos` as part of our CI](https://github.com/dealii/dealii/blob/16484460f0d2a2a02dcb45acb31b2715277dc3cf/.github/workflows/indent.yml#L72-L83), and have a corresponding configuration file [`.typos.toml`](https://github.com/dealii/dealii/blob/master/.typos.toml) in the root folder of the repository.
> 
> It would be great to use the same spell checker for `pre-commit`. Would it be possible to add a hook for `typos` instead of `codespell`?

Yes, typos has a pre-commit hook too, added in the newest commit with the same settings as in `indent.yml` (and removed it there)

jpthiele commented:
> I like this! 👍
> 
> Could you please also add a changelog entry for this new feature?
> 
> And please also add some instructions on how to set-up `pre-commit` with all its dependencies. I guess a new entry in this [Wiki section](https://github.com/dealii/dealii/wiki#developingcontributing) would be a good place.

Sure I'll work on that this afternoon.

marcfehling commented:
> Yes, typos has a pre-commit hook too, added in the newest commit with the same settings as in indent.yml (and removed it there)

I'm okay with moving the `typos` check from the `indent` workflow to `pre-commit`.

For github-actions, we are using `dependabot` to keep all dependencies up to date. `typos` has frequent releases (multiple per month), and with `dependabot` we could keep it up-to-date easily.

Do you know if there is an easy way to keep these dependencies also up-to-date in `.pre-commit-config.yaml`? If not, then I would vote to keep the `typos` check in the `indent` workflow in addition to the one in the `pre-commit` workflow.

marcfehling commented on .github/workflows/pre-commit.yml:
```diff
@@ -0,0 +1,15 @@
+name: pre-commit
+
+on:
+  pull_request:
```
```suggestion
on: [push, pull_request]
```
I believe it would be beneficial to run it also on every push, similar to what we are doing for the `indent` and `tidy` workflows.

jpthiele replied:
On push has the problem that the workflow will fail due to the 'no accidental commit to master' hook. 
But maybe we could skip said hook in case of a push?

jpthiele commented:
> > Yes, typos has a pre-commit hook too, added in the newest commit with the same settings as in indent.yml (and removed it there)
> 
> I'm okay with moving the `typos` check from the `indent` workflow to `pre-commit`.
> 
> For github-actions, we are using `dependabot` to keep all dependencies up to date. `typos` has frequent releases (multiple per month), and with `dependabot` we could keep it up-to-date easily.
> 
> Do you know if there is an easy way to keep these dependencies also up-to-date in `.pre-commit-config.yaml`? If there is no easy way, then I would vote to keep the `typos` check in the `indent` workflow for now in addition to the one in the `pre-commit` workflow.

pre-commit has the command `pre-commit autoupdate`to bump dependencies.
I found [this blog post](https://browniebroke.com/blog/gh-action-pre-commit-autoupdate/) which built a timed CI yaml similar to what dependabot does.

jpthiele commented:
Whitespace check for changelog entries was missing. Applied when added the changelog entry.
I'm happy to squash and rebase once a set of hooks is agreed upon.

jpthiele commented:
> And please also add some instructions on how to set-up `pre-commit` with all its dependencies. I guess a new entry in this [Wiki section](https://github.com/dealii/dealii/wiki#developingcontributing) would be a good place.

I don't think I can add changes there. I certainly can't open a PR, so I cloned the wiki and made a new repository in my namespace with a `precommit` branch. 
For the changes to the current state of the wiki see [here](https://github.com/jpthiele/dealii-Wiki/compare/master...precommit)

marcfehling commented:
> I don't think I can add changes there. I certainly can't open a PR, so I cloned the wiki and made a new repository in my namespace with a precommit branch.

You are correct. I remember now that we disabled a general write access to the wiki to protect it from malicious changes. We hope that github at some point allows to open pull requests against the wiki pages (see [discussion](https://github.com/orgs/community/discussions/9473)).

> For the changes to the current state of the wiki see [here](https://github.com/jpthiele/dealii-Wiki/compare/master...precommit)

Thank you for writing the entry! I think the text is good, and would go ahead and embed it in our wiki.

marcfehling commented:
> pre-commit has the command pre-commit autoupdateto bump dependencies.
I found [this blog post](https://browniebroke.com/blog/gh-action-pre-commit-autoupdate/) which built a timed CI yaml similar to what dependabot does.

I think this is a reasonable solution.

jpthiele commented on .github/workflows/indent.yml:
```diff
@@ -68,16 +68,3 @@ jobs:
         name: doxygen_documentation
         path: build/doc/doxygen
-
-  typos:
-    # check for typos
-
-    name: typos
-    runs-on: [ubuntu-24.04]
-
-    steps:
-    - uses: actions/checkout@v4
-    - uses: crate-ci/typos@v1.27.3
-      with:
-        files: doc examples include source tests
-        config: ./.typos.toml
```
Yes they are opt-in locally and will only run on the changeset/git stage when calling `git commit` once activated.

But since the config defines some useful and sensible checks the CI job runs the checks on the full codebase just to be sure. 
So it is kind of the other way around. With the config and CI YAML we have a sensible CI check, 
which can easily be automated locally as well if you want to avoid getting "CI job failed" notifications minutes after updating your PR.

bangerth replied:
OK. The point I wanted to make is that we *still* have to run these checks on the CI, even if they are available as pre-commit checks. Do I understand right that that they are still running?

jpthiele replied:
Ah got it.
Yes, a run of typos over the whole codebase is still performed as part of the CI.
But in `pre-commit` and no longer in `indent`.
